### PR TITLE
feat(schema-engine): `CHECK` and `EXCLUDE` constraints stopgap for Postgres

### PR DESCRIPTION
**Pull request by jkomyno**
- Contributes to https://github.com/prisma/schema-team/issues/384
- Contributes to https://github.com/prisma/schema-team/issues/395
- Contributes to https://github.com/prisma/prisma/issues/17515.

Docs:
- [Check Constraints](https://www.notion.so/prismaio/Indexes-Constraints-Check-constraints-PostgreSQL-cde0bee25f6343d8bbd0f7e84932e808)
- [Exclusion Constraints](https://www.notion.so/prismaio/PostgreSQL-Exclusion-Constraints-fb2ecc44f773463f908d3d0e2d737271)

Note: the stopgap implementation for `Check` constraints is exactly the same as the one for `Exclusion` constraints for Postgres. Hence why they were implemented at the same time.

In terms of introspection warning codes:
- `CHECK`: 33
- `EXCLUDE`: 34

## TODOs

- [x] add a comment at the top of the model explaining that `EXCLUSION` and `CHECK` constraints couldn't be added as they're not supported 
- [x] fix https://github.com/prisma/prisma/issues/17515

### CODE REVIEW

File: schema-engine/sql-schema-describer/src/postgres/constraints_query.sql
Comment (jkomyno) [inline schema-engine/sql-schema-describer/src/postgres/constraints_query.sql:15]: This excludes:
- `p`: `PRIMARY KEY` constraints
- `u`: `UNIQUE` constraints, which are retrieved via indexes
- `f`: `FOREIGN KEY` constraints

File: schema-engine/sql-introspection-tests/tests/postgres/mod.rs
Comment (janpio) [inline schema-engine/sql-introspection-tests/tests/postgres/mod.rs:269]: Nice, that will be interesting to see how it deals with all the ones we should have in Introspection-CI 👍

File: schema-engine/sql-schema-describer/src/postgres/constraints_query.sql
Comment (tomhoule) [reply schema-engine/sql-schema-describer/src/postgres/constraints_query.sql:15]: Is there an obstacle to querying this in the same query as foreign keys (I think it's the same tables as the existing foreign keys query)? That would save a roundtrip to the database for every call to describe, and we do call describe a lot in migrate.

Comment (jkomyno) [reply schema-engine/sql-schema-describer/src/postgres/constraints_query.sql:15]: The major obstacle is that foreign keys and check/exclusion constraints require different info, and use very different JOINs.

How about:

```sql
WITH cte_constraints AS (
	SELECT
		constr.oid AS constraint_id,
		constr.conname AS constraint_name,
		constr.contype AS constraint_type,
		pg_get_constraintdef(constr.oid) AS constraint_definition
	FROM pg_constraint constr
	JOIN pg_class AS tableinfo
		ON tableinfo.oid = constr.conrelid
	JOIN pg_namespace AS ns
		ON ns.oid = tableinfo.relnamespace
    WHERE
      ns.nspname = ANY (ARRAY['public']) -- ANY $1
      AND constr.contype IN ('f', 'x', 'c')
)
SELECT json_build_object(
	'check',
		(
	        SELECT json_agg(json_build_object(
				'constraint_name', constr.conname,
				'constraint_definition', pg_get_constraintdef(constr.oid),
				'is_deferrable', constr.condeferrable,
				'is_deferred', constr.condeferred
	        ))
	        FROM cte_constraints ctec
	        JOIN pg_constraint constr
	        	ON constr.oid = ctec.constraint_id
	        JOIN pg_class
	        	ON constr.conrelid = pg_class.oid
	        WHERE contype = 'c'
		),
	'foreign_keys',
		(
			SELECT json_agg(json_build_object(
				'con_id', con.oid,
				'child_column', att2.attname,
				'parent_table', cl.relname,
				'parent_column', att.attname,
				'confdeltype', con.confdeltype,
				'confupdtype', con.confupdtype,
				'referenced_schema_name', rel_ns.nspname,
				'constraint_name', conname,
				'child', child,
				'parent', parent,
				'table_name', table_name,
				'namespace', namespace
			) ORDER BY con.namespace, table_name, conname, con.oid, con.colidx)
			FROM (
				SELECT
                        ns.nspname AS "namespace",
                        con1.contype,
                        unnest(con1.conkey)                AS "parent",
                        unnest(con1.confkey)                AS "child",
                        cl.relname                          AS table_name,
                        ns.nspname                          AS schema_name,
                        generate_subscripts(con1.conkey, 1) AS colidx,
                        con1.oid,
                        con1.confrelid,
                        con1.conrelid,
                        con1.conname,
                        con1.confdeltype,
                        con1.confupdtype
		        FROM cte_constraints ctec
		        JOIN pg_constraint con1
		        	ON con1.oid = ctec.constraint_id
		        JOIN pg_class cl
		        	ON con1.conrelid = cl.oid
		        JOIN pg_namespace ns
		        	ON cl.relnamespace = ns.oid
                WHERE con1.contype IN ('f', 'x', 'c')
                ORDER BY colidx
			) con
                    JOIN pg_attribute att on att.attrelid = con.confrelid and att.attnum = con.child
                    JOIN pg_class cl on cl.oid = con.confrelid
                    JOIN pg_attribute att2 on att2.attrelid = con.conrelid and att2.attnum = con.parent
                    JOIN pg_class rel_cl on con.confrelid = rel_cl.oid
                    JOIN pg_namespace rel_ns on rel_cl.relnamespace = rel_ns.oid
           )
	) as constraints
```

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (janpio) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:619]: Are these messages usually in plural even when triggered for one?

Comment (jkomyno) [reply schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs]: Yes, as `affected` is an array, we don't usually check whether such array has 1 element or more.
We only know that it is definitely not empty.

Example:

```
        [
          {
            "code": 32,
            "message": "These are exclusion constraints, which are not yet fully supported. Read more: https://pris.ly/d/postgres-exclusion-constraints",
            "affected": [
              {
                "name": "room_reservation_room_id_tstzrange_excl",
                "definition": "EXCLUDE USING gist (room_id WITH =, tstzrange(reserved_at, reserved_until) WITH &&) WHERE ((NOT canceled))"
              }
            ]
          }
        ]
``` 

Note: I have noticed that the original warning message didn't make full sense (I originally planned on including the table referenced by the constraints, but I had troubles retrieving it an is not that relevant / useful for the stopgap)

File: schema-engine/sql-introspection-tests/tests/postgres/mod.rs
Comment (janpio) [inline schema-engine/sql-introspection-tests/tests/postgres/mod.rs:317]: Shouldn't this model have a comment that it uses an unsupported exclude constraint?

File: schema-engine/sql-introspection-tests/tests/postgres/mod.rs
Comment (janpio) [inline schema-engine/sql-introspection-tests/tests/postgres/mod.rs:326]: Again the question:
Is this how warnings are usually formulated?

File: schema-engine/sql-introspection-tests/tests/postgres/mod.rs
Comment (janpio) [inline schema-engine/sql-introspection-tests/tests/postgres/mod.rs:265]: Again the question:
Is this how warnings are usually formulated?

File: schema-engine/sql-schema-describer/src/postgres/constraints_query.sql
Comment (tomhoule) [reply schema-engine/sql-schema-describer/src/postgres/constraints_query.sql:15]: > This type of queries requires rewriting parts of the logic for get_foreign_keys() and get_constraints() for Postgres

get_constraints is new so it's fine, and get_foreign_keys should not get more complicated, so I think that's fine :+1:

> The same structure for this query may not be transposable to MySQL and other databases with weaker json-aggregation support

I also think that's fine, we'll deal with that when we get there, the describers are completely separated by database and we don't make effort to sync them in that way.

---

I like the approach and I think that should work, and that addresses the extra roundtrips, let's give it a try!

---

I haven't read the query in very exhaustive details, but in the foreign keys query, it looks like we're joining twice on pg_class with the same condition (lines with the aliases `rel_cl` and `cl`).

---

re: separating the foreign keys key and the check key because different shape: again I haven't looked super deep, but it feels like they could be the same rows, with nullable columns depending on `contype` — the joins in the foreign keys query would have to become outer joins, but beyond that I'm not sure you'd have to change much. This is an impression that we could maybe do simpler, not a strongly held opinion, so please feel free to ignore.

File: schema-engine/sql-schema-describer/src/postgres.rs
Comment (eviefp) [inline schema-engine/sql-schema-describer/src/postgres.rs:720]: Why not just `.clone()` them here if this is the only place we're using those newly introduced variables?

**Review by eviefp — APPROVED** [review]
This looks good to me, but I wonder if @tomhoule or anyone else could confirm that they had at least a quick look as well.

There's one nitpick comment below.

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (tomhoule) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:608]: Is this TODO still correct? There's a comment on the models in the tests.

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (tomhoule) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:368]: Should that commented out code be there? Also, since it's all doc comments, it's going to show up in the rustdoc output for the `name` field.

Comment (tomhoule) [reply schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:368]: (same in CheckConstraint right above)

File: schema-engine/sql-schema-describer/src/postgres.rs
Comment (tomhoule) [inline schema-engine/sql-schema-describer/src/postgres.rs:1158]: I don't think we need the `IndexMap` at all here: if you use the table id for the same purpose as `constraint_key` in this function, you can mutate the `TableProperties` for the table inside the schema directly. That would be less code and avoids the memory churn with IndexMap. And indexing a `Vec` is cheaper than hash lookups an `IndexMap`.

**Review by tomhoule — CHANGES_REQUESTED** [review]
Nice! A few minor things but this is looking good, with the other PR that addresses the extra roundtrip (will have a look there next).

File: schema-engine/connectors/sql-schema-connector/src/datamodel_calculator/context/flavour/postgresql.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/datamodel_calculator/context/flavour/postgresql.rs:29]: I'd use something else here, e.g. the `ModelAndConstraint` which we already have in place.

File: schema-engine/connectors/sql-schema-connector/src/datamodel_calculator/context/flavour/postgresql.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/datamodel_calculator/context/flavour/postgresql.rs:37]: Same here, please reuse `ModelAndConstraint`.

File: schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs:51]: Is this needed just yet?

File: schema-engine/connectors/sql-schema-connector/src/rendering/models.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/rendering/models.rs:61]: We in general haven't been adding docs to the PSL, only warnings.

File: schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs:56]: Same here, not yet in the pair this stuff...

File: schema-engine/connectors/sql-schema-connector/src/rendering/models.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/rendering/models.rs:67]: Same here, we don't need to add it to the rendering really...

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:350]: We should use the `ModelAndConstraint` instead of this.

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:364]: We should use the `ModelAndConstraint` instead of this.

File: schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/warnings/generators.rs:613]: 31 is taken already, use 33 here.

File: schema-engine/sql-schema-describer/src/lib.rs
Comment (pimeys) [inline schema-engine/sql-schema-describer/src/lib.rs:794]: These two structs are identical. Do we really need both of them?

File: schema-engine/sql-schema-describer/src/postgres.rs
Comment (pimeys) [reply schema-engine/sql-schema-describer/src/postgres.rs:720]: Yeah, we don't need `_cloned` variables, just call `.clone()` directly where you need it.

File: schema-engine/sql-schema-describer/src/postgres.rs
Comment (pimeys) [inline schema-engine/sql-schema-describer/src/postgres.rs:1171]: We don't need a lambda for this. It's really hard to read, and you can just replace it with much simpler check (if you just use a standard BTreeMap):

```rust
constraints_map.entry(&constraint_key).and_modify(|curr| curr |= flag).or_insert(flag)
```

Also having closure parameters by-value, such as `(String, String)` is not really a good idea. Remvove the whole thing, and just use the `entry/and_modify/or_insert` trick in the two places where you call this.

File: schema-engine/sql-schema-describer/src/postgres/constraints_query.sql
Comment (pimeys) [reply schema-engine/sql-schema-describer/src/postgres/constraints_query.sql:15]: I agree we could load these together with foreign keys.

**Review by pimeys — CHANGES_REQUESTED** [review]

File: schema-engine/connectors/sql-schema-connector/src/rendering/models.rs
Comment (janpio) [reply schema-engine/connectors/sql-schema-connector/src/rendering/models.rs:61]: The goal was to render both warnings to the CLI and add comments to the Prisma Schema.

File: schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs:52]: This will repeat the comment in re-introspection. You have to check that `self.previous.is_none()` to prevent that.

File: schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/introspection_pair/model.rs:57]: This will repeat the comment in re-introspection. You have to check that `self.previous.is_none()` to prevent that.

**Review by tomhoule — APPROVED** [review]

File: schema-engine/sql-introspection-tests/tests/postgres/mod.rs
Comment (pimeys) [inline schema-engine/sql-introspection-tests/tests/postgres/mod.rs:529]: Umm, this still needs to be in `constraints.rs`

File: schema-engine/sql-schema-describer/src/lib.rs
Comment (pimeys) [inline schema-engine/sql-schema-describer/src/lib.rs:499]: no need for this anymore

File: schema-engine/sql-schema-describer/src/lib.rs
Comment (pimeys) [inline schema-engine/sql-schema-describer/src/lib.rs:500]: nor this

File: schema-engine/sql-schema-describer/src/postgres/indexes_query.sql
Comment (pimeys) [inline schema-engine/sql-schema-describer/src/postgres/indexes_query.sql:15]: Could we actually do exclusions here instead?

**Comment by pimeys** [comment]
you also want to rebase with main to get rid of flaky vitess

**Review by janpio — CHANGES_REQUESTED** [review]
Ensure this does not get merged in 4.13.0 accidentally

**Comment by tomhoule** [comment]
yep lots of changes since yesterday too, I'll do another review before we can merge

**Review by pimeys — APPROVED** [review]

**Review by tomhoule — APPROVED** [review]
lgtm :+1:

**Comment by jkomyno** [comment]
> Ensure this does not get merged in 4.13.0 accidentally

Could you please remove "requested changes" now? It's blocking the merge :)

**Pull request merged**
